### ISSUE TITLE
Automatic quoting identifiers inconsistent and probably not desirable.

### ISSUE THREAD

**Issue by pamtbaau** [sqlitebrowser/sqlitebrowser#1280]
### Details for the issue
I was searching the web for Sql standards with respect to quoting identifiers. Come across [this summary](https://stackoverflow.com/a/10574031/4137032), read the relevant page on [sqlite.org](https://sqlite.org/lang_keywords.html) and started to do some tests on sqlitebrowser.

SQL Standard and SQLite are on par wrt. quoting identifiers. [sqlite.org](https://sqlite.org/lang_keywords.html) states:
- 'keyword' : A keyword in single quotes is a string literal.
- "keyword": A keyword in double-quotes is an identifier.
- [keyword]: A keyword enclosed in square brackets is an identifier. This is not standard SQL. This quoting mechanism is used by MS Access and SQL Server and is included in SQLite for compatibility.
- \`keyword\`: A keyword enclosed in grave accents (ASCII code 96) is an identifier. This is not standard SQL. This quoting mechanism is used by MySQL and is included in SQLite for compatibility.

### TL;DR: Results of tests:
- sqlitebrowser does not adhere to SQLite documentation.
- sqlitebrowser does not respect explicitely added quotes by the user:
  - it may add extra quotes
  - it may remove/replace quotes.

This happens inconsistently between tables created and modified using Create Table resp. Modify Table, manually created tables and export of schema.

### Suggestion:
- sqlitebrowser should not add grave accents automatically.
  It's not according SQL Standards and the SQLite documentation.
- sqlitebrowser should not touch quotes explicitely added or not added by the user.
  The explicit use or non use of quotes is a choice of the user.

And here are the results: (yes it is a long read):
### Creating tables using Create Table wizard

1. No Quotes
Copy Create statement: CREATE TABLE \`noQuotes\` ( \`noQuotes\` INTEGER )
2. With Quotes
Copy Create statement: CREATE TABLE \`"withQuotes"\` ( \`"withQuotes"\` INTEGER )
3. With Single Quotes
Copy Create statement: CREATE TABLE \`'withSingleQuotes'\` ( \`'withSingleQuotes'\` INTEGER )
4. With Grave Quotes
Copy Create statement: CREATE TABLE \`\`\`withGraveQuotes\`\`\` ( \`\`\`withGraveQuotes\`\`\` INTEGER )

> Note: Every identifier is surrounded with extra grave quotes.

### Export tables

BEGIN TRANSACTION;
CREATE TABLE IF NOT EXISTS \`noQuotes\` ( \`noQuotes\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`"withQuotes"\` ( \`"withQuotes"\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`'withSingleQuotes'\` ( \`'withSingleQuotes'\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`\`\`withGraveQuotes\`\`\` ( \`\`\`withGraveQuotes\`\`\`	INTEGER );
COMMIT;

> Note: Same quotations as above.

### Modify tables using Modify Table wizard

1. No Quotes
Copy Create statement: CREATE TABLE "modifiedNoQuotes" ( \`modifiedNoQuotes\` INTEGER )
2. With Quotes
Copy Create statement: CREATE TABLE """modifiedWithQuotes""" ( \`"modifiedWithQuotes"\` INTEGER )
3. With Single Quotes
Copy Create statement: CREATE TABLE "'modifiedWithSingleQuotes'" ( \`'modifiedWithSingleQuotes'\` INTEGER )
4. With Grave Quotes
Copy Create statement: CREATE TABLE "\`modifiedWithGraveQuotes\`" ( \`\`\`modifiedWithGraveQuotes\`\`\` INTEGER )

> Note: Every table identifier is surrounded with extra double quotes, while columns are surrounded with extra grave quotes.

### Export modified tables

BEGIN TRANSACTION;
CREATE TABLE IF NOT EXISTS \`noQuotes\` ( \`noQuotes\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`"withQuotes"\` ( \`"withQuotes"\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`'withSingleQuotes'\` ( \`'withSingleQuotes'\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`\`\`withGraveQuotes\`\`\` ( \`\`\`withGraveQuotes\`\`\`	INTEGER );
COMMIT;

> Note: Same as previous export

### Created tables manualy

1. No Quotes
Copy Create statement: CREATE TABLE noQuotes ( noQuotes INTEGER )
2. With Quotes
Copy Create statement: CREATE TABLE "withQuotes" ( "withQuotes" INTEGER )
3. With Single Quotes
Copy Create statement: CREATE TABLE 'withSingleQuotes' ( 'withSingleQuotes' INTEGER )
4. With Grave Quotes
Copy Create statement: CREATE TABLE \`withGraveQuotes\` ( \`withGraveQuotes\` INTEGER )

> Note: This is how it should be. No change of user code.

### Export manually created tables

BEGIN TRANSACTION;
CREATE TABLE IF NOT EXISTS \`noQuotes\` ( \`noQuotes\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`"withQuotes"\` ( \`"withQuotes"\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`'withSingleQuotes'\` ( \`'withSingleQuotes'\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`\`\`withGraveQuotes\`\`\` ( \`\`\`withGraveQuotes\`\`\`	INTEGER );
COMMIT;

> Note: Same as previous export.

### Modify manually created tables with Modify Table

1. No Quotes
Copy Create statement: CREATE TABLE "modifiedNoQuotes" ( \`modifiedNoQuotes\` INTEGER )
2. With Quotes
Copy Create statement: CREATE TABLE "modifiedWithQuotes" ( \`modifiedWithQuotes\` INTEGER )
3. With Single Quotes
Copy Create statement: CREATE TABLE "modifiedWithSingleQuotes" ( \`modifiedWithSingleQuotes\` INTEGER )
4. With Grave Quotes
Copy Create statement: CREATE TABLE "modifiedWithGraveQuotes" ( \`modifiedWithGraveQuotes\` INTEGER )

> Note: All quotes removed. Every table identifier is surrounded with double quotes, while columns are surrounded with grave quotes.

### Export modified manually created tables

BEGIN TRANSACTION;
CREATE TABLE IF NOT EXISTS \`noQuotes\` ( \`noQuotes\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`"withQuotes"\` ( \`"withQuotes"\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`'withSingleQuotes'\` ( \`'withSingleQuotes'\`	INTEGER );
CREATE TABLE IF NOT EXISTS \`\`\`withGraveQuotes\`\`\` ( \`\`\`withGraveQuotes\`\`\`	INTEGER );
COMMIT;

> Note: Same as previous export

#### I'm opening this issue because:

- [ ] DB4S is crashing
- [x] DB4S has a bug
- [ ] DB4S needs a feature
- [ ] DB4S has another problem

#### I'm using DB4S on:

- [x] Windows: ( _version:_ 10 )
- [ ] Linux: ( _distro:_ ___ )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### I'm using DB4S version:

- [ ] 3.10.1
- [ ] 3.10.0
- [ ] 3.9.1
- [x] Other: 3.10.99 (Dec 29 2017)

#### I have also:

- [x] Tried out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [x] Searched for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by chrisjlocke** [comment]
I raised something similar a while ago, and it is by design as DB4S was aimed for mySql users.

**Comment by pamtbaau** [comment]
Using SQLite as database, calling the app `sqlitebrowser`, aka 'the DB Browser for SQLite' and enforce MySQL syntax? :innocent:

Nonetheless whichever SQL dialect is being used, according the tests, it does not seem to be implemented very consistently across modules....

I thought it might have been introduced by request #629 (Enforce grave quote for export to MySQL). I've also seen issues #845 and #683 which all address a part of the whole.

If fixing consistency would be considered, then I would suggest:
1. Stick to the SQLIte docs.
1. No automatic quoting of any kind.
1. No overriding of user's naming/quoting preference being it none, double, single, or grave quotes.

**Comment by mgrojo** [comment]
@pamtbaau Thanks for the detailed and documented report.

I think the way to go is making this a preference setting as suggested in #683. This is specially needed for the SQL export, that could be used for importing into another DBMS, but the user's custom should also be taken into account.

In my opinion, it would be strange to let the user write the preferred quotes in the table creation dialogs. The quoting of identifiers is an SQL syntax thing. The identifiers written in the form fields should be interpreted as is for the database, as currently done. It is up to SQLite to accept or not the characters used by the user.

I'd add this preference setting as a combo box in the SQL tab with title "Identifier quoting":
- None -> May this bring problems?
- SQL Standard: "Example"
- MS SQL Server compatible: [Example]
- MySQL compatible: `Example` -> Should this be the default?

If we need more space in the SQL tab, we can put the "SQL editor font" and "SQL editor size" in the same line. This would indeed look better and could be done for all the font settings.

**Comment by pamtbaau** [comment]
@mgrojo According to the [SQLite docs](https://sqlite.org/lang_keywords.html):
> If you want to use a keyword as a name, you need to quote it.

I interpret this as "There is no need to use quotes unless the identifier contains a reserved keyword. And if you do need quotes, use double quotes. (Btw. we included grave quotes and brackets for compatibility reasons only)". 

You wrote:
> ...it would be strange to let the user write the preferred quotes in the table creation dialogs.

It is up to the user to use or not use quotes in any dialog. Sqlitebrowser should, in my opinion, not enforce something onto the user which is not a requirement by SQLite itself, or SQL standard. Not in Create, Modify and Execute SQL dialogs.

A preference to indicate the user would like sqlitebrowser to enforce quoting in some way, may be a good option for specific use cases. Or the user might choose an option in the export dialog.

When following my interpretation of SQLite documentation, the preference setting would be:
1. Do not override the quotes I use myself. (default)
2. No quotes
3. Double quotes.
4. Grave.
5. Brackets.
No. 4. and 5. are not in order of preference.

**Comment by mgrojo** [comment]
The keywords are not the only reason to escape the identifiers. Some people create troublesome identifiers like these:

`CREATE TABLE "Manuel's" ("my column" INTEGER);`

SQLite will allow this, but it will be impossible to reference them using SQL without some kind of quoting. That's probably the rationale for escaping all the identifiers in the SQL code that SQLite produces, including quoting the identifiers that the user write in the table creation forms.

When the user creates the table using SQL, the used quotes are stored by SQLite itself. That's why in that case we are preserving that style. But when the SQL code has to be written by DB4S, we have to use quotes:

1. For avoiding any conflict with keywords.
2. For avoiding any syntax break by the character set used by the user, whom we don't want to require any knowledge of SQL syntax.

It is easier for a program to always use escaped identifiers, instead of trying to guess case by case when they are needed.

**Comment by pamtbaau** [comment]
@mgrojo Yes you are right. Quotes are also needed for otherwise invalid identifiers.

> When the user creates the table using SQL, the used quotes are stored by SQLite itself. That's why in that case we are preserving that style.

When I create tables manually, sqlitebrowser does not preserve my quotes (or absent of) in Modify Table and when exporting. Modify Table wraps all identifiers in double quotes and export wraps them in grave quotes.

> But when the SQL code has to be written by DB4S, we have to use quotes:

That's why I added the preference option 1. "Do not override the quotes I use myself. (default)". Whether this should be default or not depends on the expertise of the target audience of sqlitebrowser.

**Comment by sky5walk** [comment]
I am confused by this topic?
As mentioned, this is a SQLite browser which documents [quoting behavior here](http://www.sqlite.org/lang_keywords.html). My preference is `backticks` since they do not appear anywhere else in my code. And it is simple to concatenate literal strings without """.

**Comment by pamtbaau** [comment]
@sky5walk In the second alinea on the page you refer to, SQLite docs state:
> If you want to use a keyword as a name, you need to quote it.

Or, in other words, only if one really wants to use SQL(ite) reserved keywords as identifier (or if one want to use otherwise invalid identifiers) one needs to quote the identifier. Otherwise if one uses proper identifiers one doesn't need quotes.

Not using quotes also prevents issues in languages like for example Javascript where template literals are created using backticks:
```
const query = `
   SELECT id,
          name
     FROM product
    WHERE id = $id
`;
```
By the way, this topic is not about taking away the option of using backticks. It is about consistency throughout the product and providing the user the option to use whatever they like.

**Comment by chrisjlocke** [comment]
> It is about consistency throughout the product

Definitely.  As you originally raised, various parts of the program use different methods.
I'm also personally against grave accents ... but not for any great reason ... just because in other databases I create, they're apostrophes or normal double quotes.

**Comment by pamtbaau** [comment]
I wonder... is there any database around that requires every identifier to be quoted no matter what?

I don't recall I've ever needed to use quotes. Not in Ingres, Oracle, MS Access, MS SQL Server, MySQL, SQLite...

**Comment by sky5walk** [comment]
True, I only use quotes when my column names confuse the SQL parser.
Ex. `myCol(x)` or [myCol(x)], else SQLite and others fail with naked myCol(C).

**Comment by mgrojo** [comment]
@pamtbaau  The configuration of identifier quotes is already in the nightly build. The new default is the "SQL standard". Would you have some time for testing it? Sorry to say that we will not implement the option for not adding quotes in the application. It is too complicated to implement it well for the relatively low benefit. Just choose the quoting that you want the application to use.

**Comment by pamtbaau** [comment]
Did some test, but since I don't know the new targeted "quoting behavior", I cannot judge if the test results are "as designed".

### TLDR:
- The table wizard forces quotes around user input including explicit used (whichever) quotes.
  - Quotes used become part of the identifier.
  - I suppose this is by design and user is expected not to use quotes.
- Sql statements (create/select) are indifferent about quotes used as long as the identifier is valid.
  - SQL parser is indifferent about quotes, as long as the identifier is valid.
  - Quotes used do not become part of the identifier.
  - This behavior is in line with SQLIte/SQL standards.
- Export wraps every identifier with double quotes.
  - In contrast, SQLite3.exe uses quotes as used by user.


### Results
**Creating tables**
I've created tables using the table wizard and by executing sql statements. I used explicit quotes around identifiers according the name of the table in both the wizard and sql statements.

![test](https://user-images.githubusercontent.com/5120011/42146772-4b7da104-7dca-11e8-8dce-5188ae589f68.png)

Note:
- User is probable expected not to use quotes.
- Name column:
  - Wizard: Identifiers are shown as entered.
  - Sql statement: Identifiers used in Sql statement not shown.
- Schema column:
  - Wizard: Identifiers and used quotes are being wrapped in double quotes
  - Sql statement: Identifiers shown as entered in sql statement

**Performing selects:**
```
select * from noQuotesExecuteSql;       -- OK
select * from 'noQuotesExecuteSql';     -- OK
select * from "noQuotesExecuteSql";     -- OK
select * from `noQuotesExecuteSql`;     -- OK

-- Same results for singleQuotesExecuteSql, doubleQuotesExecuteSql, graveQuotesExecuteSql

select * from noQuotesWizard;           -- OK
select * from 'noQuotesWizard';         -- OK
select * from "noQuotesWizard";         -- OK
select * from `noQuotesWizard`;         -- OK

select * from singleQuotesWizard;       -- No such table
select * from 'singleQuotesWizard';     -- No such table
select * from "'singleQuotesWizard'";   -- OK
select * from `'singleQuotesWizard'`;   -- OK

select * from doubleQuotesWizard;       -- No such table
select * from "doubleQuotesWizard";     -- No such table
select * from """doubleQuotesWizard"""; -- OK
select * from `"doubleQuotesWizard"`;   -- OK

select * from graveQuotesWizard;        -- No such table
select * from `graveQuotesWizard`;      -- No such table
select * from "`graveQuotesWizard`";    -- OK
select * from '`graveQuotesWizard`';    -- OK
```
Note:
- Because Wizard takes user input literally, sql statements must use identifiers literally too.
- It seems like quotes used in `create table` statements are ignored and any type of quotes can be used in `select` statements.

**Export of schema SQLiteBrowser:**
```
BEGIN TRANSACTION;
DROP TABLE IF EXISTS "`graveQuotesWizard`";
CREATE TABLE IF NOT EXISTS "`graveQuotesWizard`" (
	"`id`"	INTEGER,
	"`name`"	TEXT
);
DROP TABLE IF EXISTS """doubleQuotesWizard""";
CREATE TABLE IF NOT EXISTS """doubleQuotesWizard""" (
	"""id"""	INTEGER,
	"""name"""	TEXT
);
DROP TABLE IF EXISTS "'singleQuotesWizard'";
CREATE TABLE IF NOT EXISTS "'singleQuotesWizard'" (
	"'id'"	INTEGER,
	"'name'"	TEXT
);
DROP TABLE IF EXISTS "noQuotesWizard";
CREATE TABLE IF NOT EXISTS "noQuotesWizard" (
	"id"	INTEGER,
	"name"	TEXT
);
DROP TABLE IF EXISTS "graveQuotesExecuteSql";
CREATE TABLE IF NOT EXISTS "graveQuotesExecuteSql" (
	"id"	INTEGER,
	"name"	TEXT
);
DROP TABLE IF EXISTS "doubleQuotesExecuteSql";
CREATE TABLE IF NOT EXISTS "doubleQuotesExecuteSql" (
	"id"	INTEGER,
	"name"	TEXT
);
DROP TABLE IF EXISTS "singleQuotesExecuteSql";
CREATE TABLE IF NOT EXISTS "singleQuotesExecuteSql" (
	"id"	INTEGER,
	"name"	TEXT
);
DROP TABLE IF EXISTS "noQuotesExecuteSql";
CREATE TABLE IF NOT EXISTS "noQuotesExecuteSql" (
	"id"	INTEGER,
	"name"	TEXT
);
COMMIT;
```

Notes:
- When using the wizard, quotes have become part of the identifier
- "double quotes" are forced.

**Export by SQLite3.exe v3.9.12:**
```
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;

CREATE TABLE noQuotesExecuteSql (
id INTEGER,
name TEXT
);
CREATE TABLE 'singleQuotesExecuteSql' (
'id' INTEGER,
'name' TEXT
);
CREATE TABLE "doubleQuotesExecuteSql" (
"id" INTEGER,
"name" TEXT
);
CREATE TABLE `graveQuotesExecuteSql` (
`id` INTEGER,
`name` TEXT
);

CREATE TABLE IF NOT EXISTS "noQuotesWizard" (
	"id"	INTEGER,
	"name"	TEXT
);
CREATE TABLE IF NOT EXISTS "'singleQuotesWizard'" (
	"'id'"	INTEGER,
	"'name'"	TEXT
);
CREATE TABLE IF NOT EXISTS """doubleQuotesWizard""" (
	"""id"""	INTEGER,
	"""name"""	TEXT
);
CREATE TABLE IF NOT EXISTS "`graveQuotesWizard`" (
	"`id`"	INTEGER,
	"`name`"	TEXT
);
COMMIT;

```
Notes:
- SQLite3.exe uses quotes as used by user.

**Comment by mgrojo** [comment]
@pamtbaau Thanks for the comprehensive testing and report. The results are as expected by the design.

> - The table wizard forces quotes around user input including explicit used (whichever) quotes.
>  - Quotes used become part of the identifier.
>  - I suppose this is by design and user is expected not to use quotes.

Yes, quoting is considered an SQL syntax feature for making possible that any entered string is considered an identifier. Entered strings in the text boxes of the wizard are considered the pure identifier, and that includes any quote character entered by the user. There's no point in adding quotes in the wizard, since the user can select in the preferences the quoting characters that will be used later by the application whenever the identifier is referenced in any generated SQL code.

>- Sql statements (create/select) are indifferent about quotes used as long as the identifier is valid.
>  - SQL parser is indifferent about quotes, as long as the identifier is valid.

Yes, both the internal DB4S parser as well as SQLite3 accepts the three configurable quote characters.

>  - Quotes used do not become part of the identifier.
>  - This behavior is in line with SQLIte/SQL standards.
>- Export wraps every identifier with double quotes.
>  - In contrast, SQLite3.exe uses quotes as used by user.

We produce the CREATE statements by ourselves for exporting, so the internal statement saved by SQLite3 in the schema does not affect us. We can see it as: SQLite3 remembers the user preference for quoting used for each table in the corresponding table schema; DB4S centralises this user preference in the saved application settings and uses it later unconditionally when it generates code by itself. User is nevertheless free to use whatever quoting, or lack of it, when writing code in the "Execute SQL" tab.

**Comment by pamtbaau** [comment]
Pity the option "none" was too difficult to implement. It is what it is... One day, I hope, I'll get over it 😄

**Comment by mgrojo** [comment]
@pamtbaau I'm going to close this as preparation for our next release. By the way, maybe you like the drag & drop options implemented in https://github.com/sqlitebrowser/sqlitebrowser/commit/f33943f4cf8977f975918d12fee8d6de723cdad5 It allows to drag names without any quoting to the SQL editor.

**Comment by pamtbaau** [comment]
I have downloaded the nightly this morning and tried the drag & drop, but I'm not quite sure what it is supposed to do... At least, I seem not to be able to drag a table from the database structure panel to the execute sql panel.... Could you give me a hint?

**Comment by mgrojo** [comment]
Do you select a table name in the DB Schema side panel, drag it and drop it in the SQL editor of the Execute SQL tab? That should do it. Have you played with the options in the context menu of the panel? Does that change anything?

**Comment by pamtbaau** [comment]
Ah... did miss something... The DB Schema side panel! Was trying drag&drop from the Database Structure panel. Yes, the drag and drop from the DB Schema side panel does work. 

I'm afraid I don't really get the value of it though. Would be nice if it would ask at dropping whether I would like a select, update, insert statement based on all selected columns...

**Comment by mgrojo** [comment]
> I'm afraid I don't really get the value of it though. Would be nice if it would ask at dropping whether I would like a select, update, insert statement based on all selected columns in the drag...

In its current form it already helps to compose queries. Something along the lines what you are describing is in my to-do list, but it won't happen soon. I have other things currently in hands.
